### ISSUE TITLE
Dynamic g-image tags are not lazy loaded

### ISSUE THREAD

**Issue by luka-bacic** [gridsome/gridsome#1295]
## Description

Fetching data from Contentful and dynamically creating `<g-image />` tags in a for loop disables lazy loading. All images created this way are loaded immediately on page load, whether they are in the viewport or not.

### Steps to reproduce

Here is a [repo](https://github.com/luka-bacic/reproduction-g-image-bug) demonstrating this bug. There are 2 pages, on the home page its demonstrating the `<g-image>` bug, and on the on the /about page I added a different [image lazy loader](https://github.com/alexjoverm/v-lazy-image) and these images are lazy loaded correctly. This may help to solve the issue with `<g-image>`

### Expected result

What should happen?

I expect the images to not load with the page unless they are in the viewport. The images should lazy load, just like they do if you create a `<g-image>` tag with a static image

### Actual result

What happened.

The images created from the fetched Contentful data are not lazy loaded. They are loaded with the inital page load, heavily impacting performance and SEO scores (in case there are a lot of images it can get really bad)

### Environment

Paste the information here as shown by `gridsome info`

System:
    OS: Linux 5.4 Ubuntu 18.04.4 LTS (Bionic Beaver)
    CPU: (8) x64 Intel(R) Core(TM) i7-4700HQ CPU @ 2.40GHz
  Binaries:
    Node: 10.19.0 - ~/.nvm/versions/node/v10.19.0/bin/node
    Yarn: 1.22.4 - /usr/bin/yarn
    npm: 6.14.5 - ~/.nvm/versions/node/v10.19.0/bin/npm
  Browsers:
    Chrome: 84.0.4147.89
    Firefox: 78.0.2
  npmPackages:
    @gridsome/source-contentful: ^0.5.3 => 0.5.3 
    gridsome: ^0.7.0 => 0.7.19 
  npmGlobalPackages:
    @gridsome/cli: 0.3.3

**Note about environment**: This happens to my colleagues who use a Macbook and Windows 10. It also happens on a production website

Note about me adding API keys in the repo: I don't care, I created this account solely for the purpose to demonstrate this bug.

**Comment by helloiamlukas** [comment]
I don't know if the `<g-image>` component is actually made for this purpose. I think it would make more sense to use the `g-image` directive in this case.

```vue
<template>
   ...
   <img
      :data-src="url"
      :data-srcset="srcset"
      :data-sizes="sizes"
      v-image
   />
   ...
</template>

<script>
import Image from "gridsome/app/directives/image";

export default {
  ...
  directives: {
    image: Image,
  },
  ...
}
</script>
```

Unfortunately this is not working correctly now if you use it without `data-srcset` and `data-sizes`. 

A change in `gridsome/app/directives/image.js` would prevent this:

```js
function loadImage(el) {
  ...
  if (srcset) el.srcset = srcset;
  if (sizes) el.sizes = sizes;
  ...
}
```

**Comment by luka-bacic** [comment]
Thank you for your reply helloiamlukas. Could you please explain why would it make more sense using the directive instead of a component? 

And if the `<g-image>` component was not meant for this use case, I think it should be mentioned in the docs - although honestly it doesn't make sense to me why it shouldn't work in this case. Outputting images dynamically from API data is not that uncommon. It is also obviously possible since the [v-lazy-image](https://github.com/alexjoverm/v-lazy-image) can do this out of the box

**Comment by helloiamlukas** [comment]
As far as the documentation states, `<g-image>` is mainly to output progressive images. Lazy loading is just a part of this. That's why I thought the directive might be a good choice in your case.

But I agree with you that `<g-image>` could also do this out of the box. Maybe @hjvedvik can join this discussion?

**Comment by luka-bacic** [comment]
I see, thank you for clarifying.

I think lazy loading dynamic images from an API would be a nice addition then - although I would kinda understand if the gridsome team disagrees. In case they disagree, I think it would be important to point that out in the docs.

For example, there is this alert stating only local images with relative URL paths will be compressed

> 📣 Only local, relative image paths will be compressed by Gridsome.

Maybe changing to something like this would be really informative

> 📣 Only local, relative image paths will be compressed **and lazy lodaded** by Gridsome.

**Comment by hjvedvik** [comment]
@helloiamlukas Yes, I agree. We should let `g-image` lazy load external images since Gridsome already has the ability to do it built-in. Right now it's only possible by passing a special object as the `src` property, which is very inconvenient. That object is created automatically for local images. But I think it would be much better if we could do something like this manually:
```html
<g-image
  placeholder="data:image/svg+xml,......." <!-- or a URL to a super small image -->
  src="https://url.to/large-image.jpg"
  srcset="https://url.to/medium-image.jpg 480w, https://url.to/large-image.jpg 800w"
  sizes="(max-width: 600px) 480px, 800px"
/>
```

**Comment by baernholdt-jf** [comment]
@hjvedvik That would be pretty awesome. And at the same time could we allow to disable srcsets? Maybe with `:srcset="false"` or `:sizes="false"`?

**Comment by tennox** [comment]
@hjvedvik Submitted a PR that makes this one work:

```vue
<g-image :src="{
        src: 'https://picsum.photos/200/300',
        size: {width: 200, height: 300},
        dataUri: 'data:image/gif;base64,R0lGODlhAQABAIABAP///wAAACwAAAAAAQABAAACAkQBADs=', // http://probablyprogramming.com/2009/03/15/the-tiniest-gif-ever
}" />
```

**Comment by tennox** [comment]
So, the above makes it easier to use `g-image`'s lazy-loading feature for e.g. external images, but the main problem pointed out in this issue is the pre-processing/compression of **local but dynamic assets**.

I stumbled into this when adding local images in a loop, as seen in the OP's repo:
https://github.com/luka-bacic/reproduction-g-image-bug/blob/master/src/components/DynamicPictures.vue#L6
(in my case even with a static range: `v-for="i in range(1, 32)"`)

My question is: is it possible to get the amazing pre-processing & responsiveness features of g-image also for dynamic src attributes?
Can someone maybe point me to how the images are pre-processed? I figured out the basic functionality of the AssetsQueue, but not why the dynamic rendered ones would not be added to the queue when the HTML is rendered at build time?

**Comment by hjvedvik** [comment]
Awesome, @tennox! I'll take a look at the PR pretty soon :)

Preprocessing with dynamic `src` with `g-image` is possible, but with some limitations. The value of the `src` attribute is converted into a `require()` statement with a custom loader that is built into Gridsome. So I've seen some use the loader manually like this: ``<g-image :src="require(`!!assets-loader!./images/${image}`)" />``. But keep in mind that webpack transforms the path into a regexp and loads every file matching that regex, not just images. So it's best to at least include a folder name before the variable part. And it can have some performance issues in development if the folder has a lot of images.

Truly dynamic `src` attributes will most likely not be possible until we have implemented on-demand image processing for Netlify etc. Which is something we are looking into btw.
